File: docker/lua/txt/share_grant.lua
local codec = require("txt.codec")
local config = require("txt.config")
local sodium = require("txt.sodium")

local M = {}
local VERSION = "\1"
local SALT_BYTES = 32
local NONCE_BYTES = sodium.NONCE_BYTES
local TAG_BYTES = sodium.TAG_BYTES
local KEY_INFO = "txt:share-grant-key:v1"
local AAD_PREFIX = "txt:share-grant:v1"
local ENVELOPE_BYTES = #VERSION + SALT_BYTES + NONCE_BYTES + TAG_BYTES

local function grant_key(salt, id_hash)
  local prk, err = codec.hmac("sha256", salt, config.get().share_grant_key)
  if not prk then
    return nil, err
  end
  return codec.hmac("sha256", prk, KEY_INFO .. id_hash .. "\1")
end

local function split_envelope(envelope)
  local salt = envelope:sub(2, 1 + SALT_BYTES)
  local nonce = envelope:sub(2 + SALT_BYTES, 1 + SALT_BYTES + NONCE_BYTES)
  local sealed = envelope:sub(2 + SALT_BYTES + NONCE_BYTES)
  return salt, nonce, sealed
end

function M.encrypt(id_hash, object_path)
  local salt = codec.random(SALT_BYTES)
  local nonce = codec.random(NONCE_BYTES)
  local key, err = grant_key(salt, id_hash)
  if not key then
    return nil, err
  end
  local sealed
  sealed, err = sodium.seal(key, nonce, object_path, AAD_PREFIX .. id_hash)
  if not sealed then
    return nil, err
  end
  return codec.base64url_encode(VERSION .. salt .. nonce .. sealed)
end

function M.decrypt(id_hash, grant)
  local envelope = codec.base64url_decode(grant)
  if not envelope or #envelope <= ENVELOPE_BYTES or envelope:sub(1, 1) ~= VERSION then
    return nil, "malformed grant"
  end
  local salt, nonce, sealed = split_envelope(envelope)
  local key, err = grant_key(salt, id_hash)
  if not key then
    return nil, err
  end
  return sodium.open(key, nonce, sealed, AAD_PREFIX .. id_hash)
end

return M

File: worker/shareGrant.ts
// docs/crypto.md §"Share grant envelope": how the exact R2 object path
// travels from the Worker to a share URL and back, without D1 ever storing
// it (docs/sharing.md §1). Only the Worker, holding SHARE_GRANT_KEY, can
// produce or open a grant -- this is genuinely Worker-side crypto, unlike
// the Blob Format (ui/src/crypto/), which the Worker never needs: the
// Worker holds SHARE_GRANT_KEY itself, so it's the only party able to do
// anything with a grant either way.
const VERSION = 0x01;
const SALT_LEN = 32;
const NONCE_LEN = 12;
const TAG_LEN = 16;
const MIN_LEN = 1 + SALT_LEN + NONCE_LEN + TAG_LEN; // 61

const KEY_INFO_PREFIX = new TextEncoder().encode("txt:share-grant-key:v1");
const AD_PREFIX = new TextEncoder().encode("txt:share-grant:v1");

function concat(...parts: Uint8Array[]): Uint8Array {
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

// Binds the derived key (via HKDF's `info`) and the AEAD additional data to
// this specific share's `idHash` -- decrypting under a different share's
// idHash fails the tag check, so a grant can't be replayed against another
// share (docs/crypto.md).
async function deriveKey(
  shareGrantKey: Uint8Array,
  salt: Uint8Array,
  idHash: Uint8Array,
): Promise<CryptoKey> {
  const ikmKey = await crypto.subtle.importKey("raw", shareGrantKey, "HKDF", false, [
    "deriveKey",
  ]);
  return crypto.subtle.deriveKey(
    { name: "HKDF", hash: "SHA-256", salt, info: concat(KEY_INFO_PREFIX, idHash) },
    ikmKey,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"],
  );
}

/** Seals `objectPath` into a grant envelope (docs/crypto.md), given the
 * share's `idHash = SHA-256(raw share_id)` and the Worker's
 * `shareGrantKey`. Returns the raw envelope bytes -- base64url-encoding
 * for transport is the caller's job. */
export async function sealGrant(
  objectPath: string,
  idHash: Uint8Array,
  shareGrantKey: Uint8Array,
): Promise<Uint8Array> {
  const salt = crypto.getRandomValues(new Uint8Array(SALT_LEN));
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_LEN));
  const key = await deriveKey(shareGrantKey, salt, idHash);
  const ad = concat(AD_PREFIX, idHash);
  const sealed = new Uint8Array(
    await crypto.subtle.encrypt(
      { name: "AES-GCM", iv: nonce, additionalData: ad, tagLength: TAG_LEN * 8 },
      key,
      new TextEncoder().encode(objectPath),
    ),
  );
  return concat(new Uint8Array([VERSION]), salt, nonce, sealed);
}

/** Opens a grant envelope, given the same `idHash` and `shareGrantKey` used
 * to seal it. Throws if the envelope is malformed, the version is
 * unsupported, or the AEAD tag doesn't verify (including when `idHash`
 * doesn't match the one the grant was sealed for). */
export async function openGrant(
  envelope: Uint8Array,
  idHash: Uint8Array,
  shareGrantKey: Uint8Array,
): Promise<string> {
  if (envelope.length < MIN_LEN) {
    throw new Error(`grant too short: ${envelope.length} < ${MIN_LEN}`);
  }
  if (envelope[0] !== VERSION) {
    throw new Error(`unsupported grant version: ${envelope[0]}`);
  }
  const saltStart = 1;
  const nonceStart = saltStart + SALT_LEN;
  const sealedStart = nonceStart + NONCE_LEN;
  const salt = envelope.slice(saltStart, nonceStart);
  const nonce = envelope.slice(nonceStart, sealedStart);
  const sealed = envelope.slice(sealedStart);
  const key = await deriveKey(shareGrantKey, salt, idHash);
  const ad = concat(AD_PREFIX, idHash);
  try {
    const plaintext = await crypto.subtle.decrypt(
      { name: "AES-GCM", iv: nonce, additionalData: ad, tagLength: TAG_LEN * 8 },
      key,
      sealed,
    );
    return new TextDecoder().decode(plaintext);
  } catch {
    // crypto.subtle.decrypt's own error carries no detail (by design, to
    // avoid oracle attacks); reject uniformly rather than let a raw
    // OperationError leak through with a misleading stack.
    throw new Error("grant AEAD tag verification failed");
  }
}
